packages/block-editor/src/components/block-list/block.native.js

```diff
@@ -187,18 +258,33 @@ export default compose( [
 		const icon = blockType.icon;
 		const getAccessibilityLabelExtra = blockType.__experimentalGetAccessibilityLabel;
 
-		const selectedBlock = getSelectedBlock();
-		const parentId = getBlockRootClientId( clientId );
-		const parentBlock = getBlock( parentId );
-
-		const isMediaText = selectedBlock && selectedBlock.name === 'core/media-text';
-		const isMediaTextParent = parentBlock && parentBlock.name === 'core/media-text';
+		const parents = getBlockParents( clientId, true );
+		const parentId = parents[ 0 ] || '';
 
 		const rootBlockId = getBlockHierarchyRootClientId( clientId );
 		const rootBlock = getBlock( rootBlockId );
 		const hasRootInnerBlocks = rootBlock.innerBlocks.length !== 0;
 
-		const showFloatingToolbar = isSelected && hasRootInnerBlocks && ! isMediaText && ! isMediaTextParent;
+		const showFloatingToolbar = isSelected && hasRootInnerBlocks;
+
+		const selectedBlockClientId = getSelectedBlockClientId();
+
+		const commonAncestor = getLowestCommonAncestorWithSelectedBlock( clientId );
+		const clientTree = [ clientId, ...parents ];
+		const commonAncestorIndex = clientTree.indexOf( commonAncestor ) - 1;
+		const firstToSelectId = commonAncestor || selectedBlockClientId ? clientTree[ commonAncestorIndex ] : rootBlockId;
+
+		const hasChildren = !! getBlockCount( clientId );
+		const hasParent = !! parents[ 0 ];
+		const isParentSelected = selectedBlockClientId && selectedBlockClientId === parents[ 0 ];
+		const isAncestorSelected = selectedBlockClientId && parents.includes( selectedBlockClientId );
+		const selectionIsNested = !! getBlockRootClientId( selectedBlockClientId );
+
+		const isDescendantSelected = selectedBlockClientId && getBlockParents( selectedBlockClientId ).includes( clientId );
+		const isTouchable = isSelected || isDescendantSelected || selectedBlockClientId === parentId || parentId === '';
+		const isDimmed = ! isSelected && selectionIsNested && ! isAncestorSelected && ! isDescendantSelected && ( firstToSelectId === clientId || rootBlockId === clientId );
+
+		const isGroupType = name === 'core/group';
```

Comment:
This is the only one line I need to use block specific group to avoid additional margin when selected group has nested groups inside.

I wonder if we have better way to recognise that case

Reply:
I think this is only needed if you want to make empty group blocks appear exactly like the children of a selected block. I think it makes more sense to display the dashed border but still align to the content, so we would not need any extra code for the margins.

The simple approach would look something like this (measurements are not exact):

![nested-groups](https://user-images.githubusercontent.com/8739/69550288-08310800-0f9b-11ea-8082-245a48e5ed28.png)

If we want to avoid the double border on the empty group inside the selected block, we can have that logic inside `GroupEdit` by only showing the placeholder if `! isAncestorSelected`. The previous example looks odd, but empty groups might be nested deeper and the placeholder would still be useful. See the two variants with a more complex layout:

![nested-groups-variants](https://user-images.githubusercontent.com/8739/69550848-07e53c80-0f9c-11ea-8878-51622bf20102.png)

@iamthomasbishop thoughts on this one?

Reply:
The last example you shared (where only the first level of children are outlined) is the cleanest and closest to what I was originally imagining. Showing outlines of _all_ children is okay on blocks that contain a couple of levels of nesting, but it will create a lot of visual clutter on blocks with deeper nesting.

Reply:
I think it's cleaner, but to clarify, it would only add borders to an empty group, so any group that contains just another group wouldn't be empty, and only the "leaf" group would have the placeholder. In summary, even the more cluttered option would show at most 3 outlines at once, regardless of how complex the hierarchy is.

Reply:
The group placeholder is one thing. I like the idea with hiding the group placeholder - not sure if the block will be still easily clickable (not to tiny to catch press event) but if that happens we can just make the dashed border of the placeholder transparent.


The issue with margins that I mentioned in initial comment is here:

Please note additional margin next to "Inner Group" of selection
<image src="https://user-images.githubusercontent.com/21242757/69621854-269d0f00-1040-11ea-840d-f39849506deb.png" width="300" />

Below it the version which using `isGroupType` to avoid this
<image src="https://user-images.githubusercontent.com/21242757/69621233-fb65f000-103e-11ea-8ae8-9099d9c97195.png" width="300" />

Reply:
I' working on get rid of `isGroupType` prop and it almost done. There is still one line to handle but I believe I will manage it somehow.

According to the placeholders that @koke mentioned I prepare a GIF about how it looks. I like the "clean" solution but I think presented one ( where we see group placeholder inside dashed border applied by navigation logic) is more consistent. Please take a look:
 
<image src="https://user-images.githubusercontent.com/21242757/69645172-06804680-1066-11ea-9a0e-8387f14008be.gif" width"=300" />

Reply:
With @iamthomasbishop and @koke we decided to keep the version which is presented on posted GIF above.